tests/test_interpreter.py

```diff
@@ -0,0 +1,156 @@
+import unittest
+from nelox.scanner import Scanner
+from nelox.parser import Parser
+from nelox.interpreter import Interpreter
+
+
+class InterpreterTest(unittest.TestCase):
+
+    def run_code(self, source):
+        scanner = Scanner(source)
+        parser = Parser(scanner)
+        expressions = parser.parse()
+        interpreter = Interpreter()
+        return interpreter.interpret(expressions)
+
+    def test_simple_arithmetic(self):
+        self.assertEqual(self.run_code("(+ 1 2)"), 3)
+        self.assertEqual(self.run_code("(* 3 4)"), 12)
+        self.assertEqual(self.run_code("(- 10 3)"), 7)
+        self.assertEqual(self.run_code("(/ 8 2)"), 4.0)
+
+    def test_set_variable(self):
+        result = self.run_code("""
+            (set x 42)
+            x
+        """)
+        self.assertEqual(result, 42)
+
+    def test_update_variable(self):
+        result = self.run_code("""
+            (set x 1)
+            (set x 99)
+            x
+        """)
+        self.assertEqual(result, 99)
+
+    def test_variable_shadowing(self):
+        result = self.run_code("""
+            (set x 10)
+            (func f (x) (+ x 1))
+            (f 5)
+        """)
+        self.assertEqual(result, 6)
+
+    def test_if_true_branch(self):
+        result = self.run_code("""
+            (if (> 5 3)
+                1
+                0)
+        """)
+        self.assertEqual(result, 1)
+
+    def test_if_false_branch(self):
+        result = self.run_code("""
+            (if (< 2 1)
+                1
+                0)
+        """)
+        self.assertEqual(result, 0)
+
+    def test_function(self):
+        result = self.run_code("""
+            (func square (x) (* x x))
+            (square 6)
+        """)
+        self.assertEqual(result, 36)
+
+    def test_lambda_simple(self):
+        result = self.run_code("""
+            ((lambda (a b) (+ a b)) 4 6)
+        """)
+        self.assertEqual(result, 10)
+
+    def test_lambda_closure(self):
+        result = self.run_code("""
+            (set x 100)
+            (set make_adder (lambda (y) (lambda (z) (+ x y z))))
+            (set add_all (make_adder 5))
+            (add_all 2)
+        """)
+        self.assertEqual(result, 107)
```

Comment:
A similar test to this is useful: a counter, where you define a function that returns a lambda with a local variable x, that is incremented and returned every time.